### PR TITLE
EINTR aware std.file read, write and copy on Posix

### PR DESCRIPTION
**Pull request by FreeSlave**
Repeat syscalls if they were interrupted by signal.

### CODE REVIEW

File: std/file.d [hunk 1]
```diff
@@ -95,6 +95,22 @@ else version (Posix)
 {
     import core.sys.posix.dirent, core.sys.posix.fcntl, core.sys.posix.sys.stat,
         core.sys.posix.sys.time, core.sys.posix.unistd, core.sys.posix.utime;
+
+    template deintr(alias func)
```
Comment (CyberShadow) [inline std/file.d]: This should probably be private.

**Comment by FreeSlave** [comment]
> Perhaps alarm + ignoring SIGALARM to check that it does at least not throw?
I'm not sure how it would help. I believe if signal is being blocked, there will be no interruption either.

I don't think there's a reliable way to reproduce syscall interruption.

**Comment by CyberShadow** [comment]
> I believe if signal is being blocked, there will be no interruption either.

I thought I remember reading that even if a signal is ignored, it will still interrupt syscalls. But I might be wrong. In any case, an empty signal handler should work, right?

**Comment by FreeSlave** [comment]
I still don't understand how you gonna ensure that signal will be delivered at the time right for syscall interruption to produce EINTR.

The only way to test deintr I think of is a separate program doing some long job and constantly getting signals during it job. E.g. https://gist.github.com/FreeSlave/c97120ebefa01f63f34bdeb202a7d26f
But for including to unittests such test is too brute force and still formally not reliable.

**Comment by CyberShadow** [comment]
> E.g. https://gist.github.com/FreeSlave/c97120ebefa01f63f34bdeb202a7d26f

That looks more complicated than it should be. Why not just copy from /dev/zero to /dev/null?

**Comment by FreeSlave** [comment]
Should work too. But the point is you can't ensure that any of delivered signals will lead to syscall return EINTR.
Also do you want to test read/write/copy against signals or just deintr itself?

**Comment by FreeSlave** [comment]
Also I've just reliazed that my gist does not even use deintr template. It's just attempt to produce EINTR errors. All writes and reads should be replaced with deintr wrappers to ensure that there will be no EINTR errors.

**Comment by CyberShadow** [comment]
Hmm. Alright.

May I ask, do you have a use case for this change, or is it purely a correctness fix? As I understand, EINTR should be avoidable on all POSIX systems D supports, either via SA_RESTART or as a property of the system (*BSD).

Seeing as how a reasonable unit test appears to be impractical, I think this can be merged as-is.

**Comment by FreeSlave** [comment]
More like correctness fix.
AFAIK D uses signals in GC implementation. Therefore in threaded application GC signals may interrupt syscalls (they can't do it in single threaded application because GC can't run at the same time as syscall).
Also I think such library functions better should be atomical.

**Comment by wilzbach** [comment]
>  I think this is worth noticing in the changelog - just in case people on obscure Posix systems are affected by this. Also it's nice to show-off our work :)

I forgot to add: I added a changelog entry to your PR - please let me know if you approve.

**Comment by FreeSlave** [comment]
Changelog is ok. EINTR is not a signal though but errno code that returned by syscalls if they are interrupted by signal.

**Comment by FreeSlave** [comment]
Actually I found that manual restarting is not required for regular file I/O.
From signal(7) on Linux:
`
If  a blocked call to one of the following interfaces is interrupted by a signal handler, then the call will be automatically restarted after the signal handler returns if the SA_RESTART flag was used; otherwise the call will
       fail with the error EINTR:

       * read(2), readv(2), write(2), writev(2), and ioctl(2) calls on "slow" devices.  A "slow" device is one where the I/O call may block for an indefinite time, for example, a terminal, pipe, or socket.  If an I/O call on a  slow
         device  has  already  transferred  some data by the time it is interrupted by a signal handler, then the call will return a success status (normally, the number of bytes transferred).  Note that a (local) disk is not a slow
         device according to this definition; I/O operations on disk devices are not interrupted by signals.

       * open(2), if it can block (e.g., when opening a FIFO; see fifo(7)).
`
It does not say about `ftruncate` and `close` though. And it may be Linux-specific.

**Comment by CyberShadow** [comment]
Yep, but then there's things like NFS. One can find a bunch of EINTR + NFS mentions online: https://www.google.com/search?q=eintr+nfs

**Comment by FreeSlave** [comment]
Can you provide links to specific articles?

**Comment by CyberShadow** [comment]
Sorry, what do you mean? Not sure if there are any official publications on the matter, but it does seem a common mention in online discussions.

**Comment by FreeSlave** [comment]
Or provide links to certain discussions.
That link to google search does not help much. For me the first and the third results are links to the list of errno codes and the second talks about sockets, not nfs. Google search results also may differ for different users.

**Comment by CyberShadow** [comment]
[4th result](https://lwn.net/Articles/365294/) mentions that close can return EINTR on NFS.
[5th result](https://lwn.net/Articles/576478/) is a detailed article discussing returning EINTR on close. Comments mention EINTR on NFS.
[6th result](https://stackoverflow.com/questions/11844284/why-isnt-write2-returning-eintr) is a Stack Overflow question about EINTR. Comment on question mentions NFS.

I'm not sure what the point of me writing the above is.

**Comment by FreeSlave** [comment]
just trying to figure out which (if any) syscalls may need to be explicitly restarted when opening/reading/writing/closing regular files.

**Comment by FreeSlave** [comment]
From linux man page for `close`:
```
       Retrying  the close() after a failure return is the wrong thing to do, since this may cause a reused file descriptor from another thread to be closed.  This can occur because the Linux kernel always releases the file descrip‐
       tor early in the close operation, freeing it for reuse; the steps that may return an error, such as flushing data to the filesystem or device, occur only later in the close operation.
```
So I guess it's better to not loop on `close`.

**Comment by wilzbach** [comment]
> Actually I found that manual restarting is not required for regular file I/O.

@FreeSlave what's the holdup here? Did you manage to find out for which syscall looping is required?

**Comment by FreeSlave** [comment]
We don't need any loops at all at least in the case of regular file operations, since they are not considered 'slow' devices by signals and are not interruptible.
I'd better close this PR since it does not change anything besides operations with regular files.

**Comment by FreeSlave** [comment]
Or leave looping for ftruncate only. It's still not clear for me if we need loop here.

**Pull request closed without merging**